### PR TITLE
HIP: Initialize device-related variables only by the singleton

### PR DESCRIPTION
**Pull request by masterleinad**
Analogous to #5442 for HIP.

### CODE REVIEW

File: core/src/HIP/Kokkos_HIP.cpp
Comment (dalg24) [inline core/src/HIP/Kokkos_HIP.cpp:71]: If we are refactoring we should just get rid of that `Impl::HIPInternalDevices` altogether

File: core/src/HIP/Kokkos_HIP.cpp
Comment (dalg24) [inline core/src/HIP/Kokkos_HIP.cpp:69]:
```suggestion
  const int hip_device_id = Impl::get_gpu(settings);
```

File: core/src/HIP/Kokkos_HIP.cpp
Comment (dalg24) [inline core/src/HIP/Kokkos_HIP.cpp:76]: This is unnecessary.  The `device_id` validation is done by Kokkos Core

File: core/src/HIP/Kokkos_HIP.cpp
Comment (dalg24) [inline core/src/HIP/Kokkos_HIP.cpp:120]:
```suggestion
  Impl::initialize_host_hip_lock_arrays();
```

File: core/src/HIP/Kokkos_HIP_Instance.cpp
Comment (dalg24) [inline core/src/HIP/Kokkos_HIP_Instance.cpp:92]: Declare them `inline` and define them inside the `HIPInternal` class definition instead?

File: core/src/HIP/Kokkos_HIP_Instance.hpp
Comment (dalg24) [inline core/src/HIP/Kokkos_HIP_Instance.hpp:124]: Readability, move it next to the other static members

**Review by dalg24 — CHANGES_REQUESTED** [review]
Note that it likely conflicts with #5448

File: core/src/HIP/Kokkos_HIP.cpp
Comment (ldh4) [inline core/src/HIP/Kokkos_HIP.cpp:73]: Seeing that `HIP::detect_device_count()` now does the same thing as this line,

```suggestion
  int hipDevCount = HIP::detect_device_count();
```

**Comment by masterleinad** [comment]
THE HIP CI builds pass. A couple of CUDA CI steps fail because the tester was running out of disk space.

**Comment by masterleinad** [comment]
Updated after #5448 was merged.

**Review by dalg24 — APPROVED** [review]
Fine by me but requiring approval of Bruno

**Review by Rombur — APPROVED** [review]
Looks fine but you need to fix the conflict.

File: core/src/HIP/Kokkos_HIP_Instance.cpp
Comment (dalg24) [reply core/src/HIP/Kokkos_HIP_Instance.cpp:92]: Ping

**Comment by masterleinad** [comment]
> Looks fine but you need to fix the conflict.

Merged with `develop` again...

**Comment by Rombur** [comment]
This can be merged. The CI failures are clearly unrelated.

**Pull request merged**
